appdev/guides/pushnotifications.rst

```diff
@@ -0,0 +1,235 @@
+
+Let's practice this step-by-step.
+
+.. note::
+    In the following example we will not implement a server. Also the communication between your app and your server is up to you. Please inform the user about the communication with your server by providing a privacy policy!
+
+Make the app ready for push notifications
+-----------------------------------------
+
+Implementing the PushClient
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+First we need to add the policy group „push-notification-client“. Your apparmor file could look like this::
+
+	{
+	    "policy_groups": [
+		"networking",
+		"push-notification-client"
+	    ],
+	    "policy_version": 16.04
+	}
+
+In the next step we need to modify the Qml parts. We need to add a pushclient component::
+
+	[…]
+
+	import Ubuntu.PushNotifications 0.1
+
+	[…]
+
+	PushClient {
+		id: pushClient
+		appId: "appname.yourname_hookname"
+		onTokenChanged: console.log("👍", pushClient.token)
+	}
+
+You need to set the correct appId! If the app name in your manifest file is appname.yourname and the name of the hook is hookname, then the appId is:  appname.yourname_hookname.
+When we now start the app, it will get a token and print this token in the logs. With clickable logs we will be able to copy this token out of the terminal. But the app is not yet ready to receive a push notification. For this we need something called a pushhelper!
+
+Implementing the push helper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The pushhelper is a part of the app which will receive all push notifications and process them before sending them to the system notification center. It will receive a json-file and must outpot another json-file in the correct format. The pushhelper is seperated from the app. So we need a new hook in the manifest. It could look like this::
+
+	{
+
+		[…]
+
+		"title": "pushclient",
+	    "hooks": {
+		"pushclient": {
+		    "apparmor": "pushclient.apparmor",
+		    "desktop":  "pushclient.desktop"
+		},
+		"push": {
+		    "apparmor": "push-apparmor.json",
+		    "push-helper": "push.json"
+		}
+	    },
+
+		[…]
+
+	}
+
+It should be clear that we now need a different apparmor file and a different executable file. The push-apparmor.json file must only contain the policy group push-notification-client and should look like this::
+
+	{
+	    "template": "ubuntu-push-helper",
+	    "policy_groups": [
+		"push-notification-client"
+	    ],
+	    "policy_version": 16.04
+	}
+
+The push.json is just for redirecting to the executable file::
+
+	{
+	    "exec": "pushexec"
+	}
+
+In our tutorial we will use python to create a simple executable which will just forward the notification without chaning anything::
+
+	#!/usr/bin/python3
+
+	import sys
+
+	f1, f2 = sys.argv[1:3]
+
+	open(f2, "w").write(open(f1).read())
+
+If we don't have done it already we also need to add this new files to the cmakelist and make the pushexec executable::
```

Comment:
Drop the start of this sentence so it becomes "We also need to add these new files..."

Reply:
Done :)